core/src/main/scala/org/bitcoins/core/script/control/ControlOperationsInterpreter.scala

```diff
@@ -22,111 +20,104 @@ sealed abstract class ControlOperationsInterpreter {
+      val stackTopOpt = program.stack.headOption
+
+      stackTopOpt match {
+        case None =>
+          logger.error("We do not have any stack elements for our OP_IF")
+          program.failExecution(ScriptErrorUnbalancedConditional)
+        case Some(stackTop) =>
+          if (isNotMinimalStackTop(stackTop, sigVersion, minimalIfEnabled)) {
+            logger.error(
+              "OP_IF argument was not minimally encoded, got: " + stackTop)
+            program.failExecution(ScriptErrorMinimalIf)
+          } else {
+            val stackTopTrue: Boolean = program.stackTopIsTrue
+            logger.debug(s"OP_IF stack top was $stackTopTrue")
+            logger.debug(s"Stack top: ${program.stack}")
+
+            program
+              .updateStackAndScript(program.stack.tail, program.script.tail)
+              .addCondition(stackTopTrue)
+          }
+      }
     } else {
-      logger.debug("OP_IF stack top was false")
-      //remove the OP_IF
-      val scriptWithoutOpIf: BinaryTree[ScriptToken] = removeFirstOpIf(
-        binaryTree)
-      program.updateStackAndScript(program.stack.tail, scriptWithoutOpIf.toList)
+      // Doesn't matter which condition we use here,
+      // just that one gets added to keep track of depth
+      program.updateScript(program.script.tail).addCondition(condition = true)
     }
   }
 
   /** Checks if the stack top is NOT minimially encoded */
   private def isNotMinimalStackTop(
-      stackTopOpt: Option[ScriptToken],
+      stackTop: ScriptToken,
       sigVersion: SignatureVersion,
       minimalIfEnabled: Boolean): Boolean = {
     //see: https://github.com/bitcoin/bitcoin/blob/528472111b4965b1a99c4bcf08ac5ec93d87f10f/src/script/interpreter.cpp#L447-L452
     //https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-August/013014.html
-    val isNotMinimal = stackTopOpt.map { stackTop =>
-      (sigVersion == SigVersionWitnessV0 && minimalIfEnabled
-      && !BitcoinScriptUtil.isMinimalToken(stackTop))
-    }
-    isNotMinimal.isDefined && isNotMinimal.get
+    (sigVersion == SigVersionWitnessV0 && minimalIfEnabled
+    && !BitcoinScriptUtil.isMinimalToken(stackTop))
   }
 
   /** If the top stack value is 0, the statements are executed. The top stack value is removed. */
   def opNotIf(
       program: ExecutionInProgressScriptProgram): StartedScriptProgram = {
     require(program.script.headOption.contains(OP_NOTIF),
             "Script top was not OP_NOTIF")
-    val minimalIfEnabled = ScriptFlagUtil.minimalIfEnabled(program.flags)
-    val sigVersion = program.txSignatureComponent.sigVersion
-    val oldStackTop = program.stack.headOption
 
-    if (isNotMinimalStackTop(oldStackTop, sigVersion, minimalIfEnabled)) {
-      //need to duplicate minimal check, we cannot accurately invert the stack
-      //top for OP_IF otherwise
-      program.failExecution(ScriptErrorMinimalIf)
+    if (program.isInExecutionBranch) {
```

Comment:
nit: You might be clever enough to figure this out:

Can we invert the stack top, and the call `OP_IF` code? `OP_NOTIF` is just the inverted version of `OP_IF` right?

Reply:
While this is certainly doable and not even hard. It turns out not to clean anything up since before passing over to `opIf`, `opNotIf` must first check `program.isInExecutionBranch` (since we don't want to invert stack top if not) as well as checking for `MinimalIf` because if it is not minimally encoded we can't invert it into something that is without avoiding a necessary ScriptError.

All of this means that the only code that gets replaced is:
```
else {
            val stackTopFalse: Boolean = program.stackTopIsFalse
            logger.debug(s"OP_IF stack top was $stackTopFalse")
            logger.debug(s"Stack top: ${program.stack}")

            program
              .updateStackAndScript(program.stack.tail, program.script.tail)
              .addCondition(stackTopFalse)
          }
```
and I don't think that replacing that with a stackTop inversion followed by a call to `opIf` is cleaner

Reply:
That being said, I was able to factor out the common code